engines/sci/engine/vm_hooks.cpp

```diff
@@ -0,0 +1,143 @@
+
+namespace Sci {
+
+/** Write here all games hooks
+ *  From this we'll build _hooksMap, which contains only relevant hooks to current game
+ *  The match is performed according to PC, script number, objName, opcode (only opcode name, as seen in ScummVM debugger),
+ *  and either:
+ *  - selector	(and then externID is -1)
+ *  - external function ID (and then selector is "")
+ */
+static const GeneralHookEntry allGamesHooks[] = {
+	// GID	 ,  PC.seg, PC.offset, script, objName, selector, externID, opcode, function
+	{GID_QFG1, {0x0018, 0x144d}, {58, "egoRuns", "changeState", -1 , "push0", &qfg1_die_after_running_on_ice}},
+	// the following is example, can be removed when we'll have a concrete usage
+	{GID_QFG1, {0x0001, 0x199e}, {0 , "egoRuns", ""           , 36 , "ret",   &qfg1_extern_example}}
+};
+
+
+VmHooks::VmHooks() {
+	// build _hooksMap
+	for (uint i = 0; i < sizeof(allGamesHooks) / sizeof(GeneralHookEntry); i++) {
+		if (allGamesHooks[i].gameId == g_sci->getGameId())		
+			_hooksMap.setVal(allGamesHooks[i].key, allGamesHooks[i].entry);
+	}
+}
+
+uint64 HookHashKey::hash() {
+	return ((uint64)segment << 32) + offset;
+}
+
+
+// solves the issue described at #9646:
+// "
+// When in room 58, and type "run", the hero will fall and his HP will decrease by 1 point. This can be repeated, but will never cause the hero to die.
+// When typing "run" the ego will be assigned with the script egoRuns.
+// egoRuns::changeState calls proc0_36 in script 0 which is deducing damage from the hero's HP.
+// This procedure returns TRUE if the hero is still alive, but the return value is never observed in egoRuns.
+// "
+// we solve that by calling the hook before executing the opcode following proc0_36 call
+// and check the return value. if the hero should die, we kill him
+void qfg1_die_after_running_on_ice(Sci::EngineState *s) {
+	uint32 return_value = s->r_acc.getOffset();		// I'm assuming the segment is irrelevant
+
+	if (return_value == 0) {
+		// 0 mean that the hero should die
+		// done according to the code at main.sc, proc0_29:
+		// 			(proc0_1 0 59 80 {Death from Overwork} 82 800 1 4)
+		PUSH(8);	// params count
+		PUSH(0);	// script 0
+		PUSH(59);	// extern 59
+
+		// the following commands should be:
+		// pushi    80
+		// lofsa{ Death from Overwork }
+		// push
+		//
+		// I didn't bother with finding a string address, and the following 2 commands just ignore the string:
+		PUSH(0);
+		PUSH32(s->xs->addr.pc);
+
+		PUSH(82);
+		PUSH(800);
+		PUSH(1);
+		PUSH(4);
+
+		// CALLE implementation, based on 'case op_calle' above, assuming s->r_rest is zero (or, at least, irrelevant)
+		int framesize = 16;
+		int temp = ((framesize >> 1) + 0 + 1);
+		StackPtr s_temp = s->xs->sp;
+		s->xs->sp -= temp;
+
+		ExecStack *xs_new = execute_method(s, 0, 1, s_temp, s->xs->objp, s->xs->sp[0].getOffset(), s->xs->sp);
+		if (xs_new) { // in case of error, keep old stack
+			s->_executionStackPosChanged = true;
+			s->xs = xs_new;
+		};
```

Comment:
I'm not sure if I like this implementation. We're implementing logic which should be in scripts into engine code.
Up to now, we have gone into great lengths to keep this logic into script patches. Why can't we do the same in this case?

Reply:
Are we talking about these 4 lines, or the general idea of this PR?

If it's these lines, do you have better suggestion for them?
If it's about the PR in general - see the discussion at https://bugs.scummvm.org/ticket/9646 for justification of the PR's approach.

Reply:
We are talking about the approach taken. I am aware of the bug itself, but I'm asking why are we mixing script logic and engine code with this approach.

Reply:
I assumed you meant regarding the approach. Strangely, GitHub I/F showed it as if you were speaking regarding the last 4 lines, which seemed weird.

Never mind that, let's talk about the approach.
Indeed, mixing script logic and engine code isn't recommended, I agree. But I think that sometimes, we have to take inferior design decisions, to solve problems.

And as far as I have understand, we are already mixing script logic and engine code to solve some bugs that are impossible/difficult to solve in the regular ways.

Just looked through the code, and copied some random examples:

kgraphics.cpp, lines 1171-1183:

	reg_t kAnimate(EngineState *s, int argc, reg_t *argv) {
		reg_t castListReference = (argc > 0) ? argv[0] : NULL_REG;
		bool cycle = (argc > 1) ? ((argv[1].toUint16()) ? true : false) : false;

		g_sci->_gfxAnimate->kernelAnimate(castListReference, cycle, argc, argv);

		// WORKAROUND: At the end of Ecoquest 1, during the credits, the game
		// doesn't call kGetEvent(), so no events are processed (e.g. window
		// focusing, window moving etc). We poll events for that scene, to
		// keep ScummVM responsive. Fixes ScummVM "freezing" during the credits,
		// bug #3101846
		if (g_sci->getGameId() == GID_ECOQUEST && s->currentRoomNumber() == 680)
			g_sci->getEventManager()->getSciEvent(kSciEventPeek);


kpathing.cpp, lines 1188-1196:

	// WORKAROUND: broken polygon in lsl1sci, room 350, after opening elevator
	// Polygon has 17 points but size is set to 19
	if ((size == 19) && g_sci->getGameId() == GID_LSL1) {
		if ((s->currentRoomNumber() == 350)
		&& (readPoint(pointList, 18) == Common::Point(108, 137))) {
			debug(1, "Applying fix for broken polygon in lsl1sci, room 350");
			size = 17;
		}
	}

kscripts.cpp, lines 291-299:

	// WORKAROUND: Bugfix for the intro speed in PQ2 version 1.002.011.
	// This is taken from the patch by NewRisingSun(NRS) / Belzorash. Global 3
	// is used for timing during the intro, and in the problematic version it's
	// initialized to 0, whereas it's 6 in other versions. Thus, we assign it
	// to 6 here, fixing the speed of the introduction. Refer to bug #3102071.
	if (g_sci->getGameId() == GID_PQ2 && script == 200 &&
		s->variables[VAR_GLOBAL][kGlobalVarSpeed].isNull()) {
		s->variables[VAR_GLOBAL][kGlobalVarSpeed] = make_reg(0, 6);
	}

Reply:
Anyway, I've got an idea that maybe will improve things, so let's put this in hold, and I will check it.

Reply:
The code you mention is out of place indeed, but most of it was added in early stages of the SCI engine development, before the script patcher was implemented. It should be moved into script patches.

Reply:
I've separated script logic from engine code.
But it comes with a price...

I had to modify opcode behaviour in `vm.cpp`, see `case op_bt`.
If it's OK, I will need to change also `op_bnt`.

(regarding `jmp` - I will make it just illegal in these patches, without touching `vm.cpp`)

That's because the hook's instructions aren't placed in the interpreter memory, but come from a "bypass" - and most critical - the PC has the same value during their execution.

Reply:
I'm working on improvement that will allow `bnt` with negative offset, and `jmp` will be supported as well.

But still, I'd like to hear your opinion regarding the last commit I've pushed.

Reply:
Well, I've fixed the issues I've previously mentioned (after separating script logic and engine code).

I think that now the code is ready.

Reply:
Found a problem.
I've debugged other issue, and noticed that some instruction was initially at `0040:01e7`, but on another debugging session it was at `003e:01e7`.
So, the `segment` might change, and I probably shouldn't check it.

I hope that we *can* count on `offset` being constant (as I saw until now), and use the combination of `script number + offset`.

Reply:
Fixed that.
Ready for a new review.